### PR TITLE
Step1

### PR DESCRIPTION
**Pull request by iamrobinhood12345**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,3 @@
-# http-server
\ No newline at end of file
+# http-server
+
+An http server using client sockets.
```
Comment (nhuntwalker) [inline README.md]: You could stand to write more than this in your README

File: src/client.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    if len(message) % 8 == 0:
+        message += '$'
+    info = socket.getaddrinfo('127.0.0.1', 6011)
+    stream_info = [i for i in info if i[1] == socket.SOCK_STREAM][0]
+    client_socket = socket.socket(*stream_info[:3])
+    client_socket.connect(stream_info[-1])
+    buffer_length = 8
+    response = ""
+    print(message.encode('utf-8'))
+    client_socket.sendall(message.encode('utf-8'))
```
Comment (nhuntwalker) [inline src/client.py]: `encode` will break this in Python 2

File: src/client.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    response_list = []
+    while True:
+        data = client_socket.recv(buffer_length).decode('utf-8')
+        response_list.append(data)
+        if len(data) < buffer_length:
+            break
+    client_socket.close()
+    response = ''.join(response_list)
```
Comment (nhuntwalker) [inline src/client.py]: This will break in Python 2

File: src/server.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    """Recieve a message from the client and echos it back."""
+    while True:
+        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM, socket.IPPROTO_TCP)
+        address = ("127.0.0.1", 6011)
+        server_socket.bind(address)
+        server_socket.listen(1)
+        conn, addr = server_socket.accept()
+        buffer_length = 8
+        message_complete = False
+        response = ""
```
Comment (nhuntwalker) [inline src/server.py]: Python 2 compatability issue in waiting.

File: src/server.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        sent = False
+        while not message_complete:
+            addition = conn.recv(buffer_length).decode('utf-8')
+            response += addition
+            if response == 'bad request':
```
Comment (nhuntwalker) [inline src/server.py]: when would `response` be a `bad request`

File: src/server.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                break
+        print(response)
+        if not sent:
+            conn.sendall(response_ok().encode('utf-8'))
+        conn.close()
+
+
+def response_ok():
+    """Return 200 OK response."""
+    response = ("HTTP/1.1 200 OK\r\n" +
```
Comment (nhuntwalker) [inline src/server.py]: Why is this string in parentheses?

File: src/server.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            response += addition
+            if response == 'bad request':
+                conn.sendall(response_error().encode('utf-8'))
+                sent = True
+                break
+            elif len(addition) < buffer_length:
+                break
+        print(response)
+        if not sent:
+            conn.sendall(response_ok().encode('utf-8'))
```
Comment (nhuntwalker) [inline src/server.py]: Python 2 compatability issues

File: src/server.py (hunk 6 above)
Comment (iamrobinhood12345) [inline src/server.py]: Getting syntax errors without it

Comment (iamrobinhood12345) [reply src/server.py]: Getting syntax error without parentheses

File: src/client.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    if len(message) % 8 == 0:
+        message += '$'
+    info = socket.getaddrinfo('127.0.0.1', 6020)
+    stream_info = [i for i in info if i[1] == socket.SOCK_STREAM][0]
+    client_socket = socket.socket(*stream_info[:3])
+    client_socket.connect(stream_info[-1])
+    buffer_length = 8
+    response = u""
+    if sys.version_info[0] == 2:
+        message = message.decode('utf-8')
```
Comment (nhuntwalker) [inline src/client.py:19]: Why do you decode this if you're going to send it *immediately*?

File: src/server.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+import socket
+import sys
+
+
+def server():
+    """Recieve a message from the client and echos it back."""
+    while True:
+        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM, socket.IPPROTO_TCP)
+        address = ("127.0.0.1", 6020)
+        server_socket.bind(address)
```
Comment (nhuntwalker) [inline src/server.py]: These two lines could be one line.

File: src/server.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+
+import socket
+import sys
+
+
+def server():
+    """Recieve a message from the client and echos it back."""
+    while True:
+        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM, socket.IPPROTO_TCP)
```
Comment (nhuntwalker) [inline src/server.py]: You don't need to create a new `socket` object *every* time

File: src/test_server.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    response = ("HTTP/1.1 200 OK\r\n"
+        "Server: Teddy Bear\r\n"
+        "Content-Type: text/html\r\n"
+        "Connection: Closed\r\n\r\n"
+        "<html>\r\n"
+        "<body>\r\n"
+        "<h1>Hello, World!</h1>\r\n"
+        "</body>\r\n"
+        "</html>")
+    assert client("GET the bl®rg woot") == response
```
Comment (nhuntwalker) [inline src/test_server.py]: This should not be returning a 200 OK response

Comment (iamrobinhood12345) [reply src/test_server.py]: Im confused about what you want for this for step 1. https://codefellows.github.io/sea-python-401d5/assignments/http_server_1.html. The assignment says to build server side functions for response_ok and response_failed. How should we determine which response to send based on the request for step 1? In step 2 we implement server handling for valid and well-formed requests (https://codefellows.github.io/sea-python-401d5/assignments/http_server_2.html), so my thinking was the requirements for requests to return server responses in step 1 were trivial.

Comment (iamrobinhood12345) [reply src/test_server.py]: Does it need to be byte string? Pytesting gives error "bytes can only contain ASCII literal characters." But I thought we are supposed to handle non-ascii for requests.

File: src/client.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    """Create the protocol for interacting with server."""
+    if len(message) % 8 == 0:
+        message += '$'
+    info = socket.getaddrinfo('127.0.0.1', 6024)
+    stream_info = [i for i in info if i[1] == socket.SOCK_STREAM][0]
+    client_socket = socket.socket(*stream_info[:3])
+    client_socket.connect(stream_info[-1])
+    buffer_length = 8
+    response = u""
+    client_socket.sendall(message.encode('utf-8'))
```
Comment (nhuntwalker) [inline src/client.py]: This is *still* going to break in Python 2. You're not accounting for the fact that a string in Python2 is a bytestring by default. If a message comes through with non-standard ascii characters like ü or ñ, then you'll get an error back like this:

```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 1: ordinal not in range(128)
```

Comment (iamrobinhood12345) [reply src/client.py]:
```python
    if sys.version_info[0] == 2:
        message = message.decode('utf-8')
```
?

**Pull request merged**